### PR TITLE
Implement meet for literal types; add tests for 'is_same_type'

### PR DESCRIPTION
**Pull request by Michael0x2a**
This pull request implements meets for literal types and adds some corresponding tests.

It also adds a test suite for the `is_same_type` method while we're at it. This test suite also lets you test each type's inherent `__eq__` and `__hash__`, since we do also use those throughout the code (especially for literal types).

### CODE REVIEW

File: mypy/test/testtypes.py [hunk 1]
```diff
@@ -846,8 +848,31 @@ def test_type_type(self) -> None:
...
+
+        self.assert_meet(lit1, lit1, lit1)
+        self.assert_meet(lit1, a, lit1)
+        self.assert_meet_uninhabited(lit1, lit3)
+        self.assert_meet_uninhabited(lit1, lit2)
+        self.assert_meet(UnionType([lit1, lit2]), lit1, lit1)
+        self.assert_meet(UnionType([lit1, lit2]), UnionType([lit2, lit3]), lit2)
+        self.assert_meet(UnionType([lit1, lit2]), UnionType([lit1, lit2]), UnionType([lit1, lit2]))
+        self.assert_meet(lit1, self.fx.anyt, lit1)
+        self.assert_meet(lit1, self.fx.o, lit1)
```
Comment (ilevkivskyi) [inline mypy/test/testtypes.py]: Maybe test these two in reverse order too? (i.e. if literal is the second one)

File: test-data/unit/check-literal.test [hunk 2]
```diff
@@ -1240,3 +1240,81 @@ indirect.Literal()
...
+def f3(x: Literal[1], y: int) -> None: pass
+def f4(x: Literal[1], y: object) -> None: pass
+def f5(x: Literal[1], y: Union[Literal[1], Literal[2]]) -> None: pass
+
+reveal_type(unify(f1))  # E: Revealed type is 'Literal[1]'
+reveal_type(unify(f2))  # E: Revealed type is 'None'
+reveal_type(unify(f3))  # E: Revealed type is 'Literal[1]'
+reveal_type(unify(f4))  # E: Revealed type is 'Literal[1]'
+reveal_type(unify(f5))  # E: Revealed type is 'Literal[1]'
+[out]
```
Comment (ilevkivskyi) [inline test-data/unit/check-literal.test]: TBH, having both these tests look a bit redundant (especially that you have very good unit tests). I would combine them into a single tests and check meeting some type one way, and some types another way.

File: mypy/test/testtypes.py (hunk 1 above)
Comment (Michael0x2a) [reply mypy/test/testtypes.py]: The implementation of `assert_meet` already does this for us, thankfully: it's defined as just:

    def assert_meet(self, s: Type, t: Type, meet: Type) -> None:
        self.assert_simple_meet(s, t, meet)
        self.assert_simple_meet(t, s, meet)

...where `assert_simple_meet` is responsible for actually doing the checking.

File: test-data/unit/check-literal.test (hunk 2 above)
Comment (Michael0x2a) [reply test-data/unit/check-literal.test]: I consolidated both tests into one and pruned out some probably-overkill `reveal_type(...)` calls. I think testing the types in both ways looks tidier though, so kept that.

**Pull request merged**